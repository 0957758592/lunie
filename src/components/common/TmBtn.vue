<template>
  <button
    class="button"
    :class="{ secondary: type === `secondary`, small: size === `small` }"
    :disabled="disabled"
  >
    {{ value }}
  </button>
</template>

<script>
export default {
  name: `TmBtn`,
  props: {
    value: {
      type: String,
      default: null
    },
    type: {
      type: String,
      default: null
    },
    size: {
      type: String,
      default: null
    },
    disabled: {
      type: Boolean,
      default: false
    }
  }
}
</script>

<style scoped>
.button {
  font-family: var(--sans);
  font-size: 14px;
  font-weight: 400;
  padding: 8px 16px;
  min-width: 100px;
  color: var(--bright);
  margin: 0;
  border-radius: 0.5rem;
  cursor: pointer;
  background: var(--primary);
  border: 2px solid var(--primary);
  transition: all 0.5s ease;
  white-space: nowrap;
  outline: none;
}

.button:hover {
  background: var(--primary-dark);
  border-color: var(--primary-dark);
}

.button:disabled {
  opacity: 0.5;
  background: var(--bc);
  border-color: var(--bc-dim);
  cursor: default;
}

.button:disabled:hover {
  background: var(--bc);
  border-color: var(--bc-dim);
}

.button.secondary {
  background: var(--primary-alpha);
}

.button.secondary.small {
  padding: 6px 10px;
  font-size: 12px;
  min-width: 0;
  border-color: var(--bc);
  color: var(--dim);
  background-color: transparent;
}

.button.secondary.small:hover {
  background-color: var(--bc-dim);
}

@media screen and (max-width: 1023px) {
  .button.secondary.small {
    padding: 10px 10px;
  }
}
</style>
