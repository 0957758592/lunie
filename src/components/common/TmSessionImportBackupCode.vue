<template>
  <SessionFrame>
    <TmFormStruct :submit="() => {}">
      <h2 class="session-title">
        Recover with backup code
      </h2>
      <div class="session-main">
        <TmFormGroup
          :error="$v.$error && $v.seed.$invalid"
          field-id="import-seed"
          field-label="Backup code"
        >
          <FieldSeed
            id="import-seed"
            :value="seed"
            placeholder="Must be exactly 24 words"
            @input="val => (seed = val)"
          />
          <TmFormMsg
            v-if="$v.seed.$error && !$v.seed.required"
            name="Seed"
            type="required"
          />
          <TmFormMsg
            v-else-if="$v.seed.$error && !$v.seed.words24"
            name="Seed"
            type="words24"
          />
        </TmFormGroup>
      </div>
      <div class="session-footer">
        <TmBtn
          value="Confirm Seed"
          :disabled="$v.seed.$invalid || $v.seed.$invalid"
          @click.native="$router.push('recover-confirm-address')"
        />
      </div>
    </TmFormStruct>
  </SessionFrame>
</template>

<script>
import { mapState } from "vuex"
import { required } from "vuelidate/lib/validators"
import TmBtn from "common/TmBtn"
import TmFormGroup from "common/TmFormGroup"
import TmFormStruct from "common/TmFormStruct"
import TmFormMsg from "common/TmFormMsg"
import FieldSeed from "common/TmFieldSeed"
import SessionFrame from "common/SessionFrame"

const words24 = param => {
  return param && param.split(` `).length === 24
}

export default {
  name: `session-import-backupcode`,
  components: {
    TmBtn,
    SessionFrame,
    FieldSeed,
    TmFormGroup,
    TmFormMsg,
    TmFormStruct
  },
  computed: {
    ...mapState([`recover`]),
    seed: {
      get() {
        return this.$store.state.recover.seed
      },
      set(value) {
        console.log(`Store value for seed is ${this.$store.state.recover.seed}`)
        this.$store.commit(`updateField`, { field: `seed`, value })
      }
    }
  },
  validations: () => ({
    seed: { required, words24 }
  })
}
</script>
