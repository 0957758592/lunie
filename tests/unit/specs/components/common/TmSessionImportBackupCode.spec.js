import Vuex from "vuex"
import Vuelidate from "vuelidate"
import { mount, createLocalVue } from "@vue/test-utils"
import TmSessionImportBackupCode from "common/TmSessionImportBackupCode"

import TmBtn from "common/TmBtn"

jest.mock(`scripts/google-analytics.js`, () => () => {})
const seed = `goose toward escape engine wheel board help torch avocado educate rose rebel rigid side aspect abandon grace admit inherit female grant pledge shine inquiry`
const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(Vuelidate)
localVue.directive(`tooltip`, () => {})
localVue.directive(`focus`, () => {})

describe(`TmSessionImportBackupCode`, () => {
  let wrapper, $store

  it(`has the expected html structure`, () => {
    const getters = {}
    $store = {
      state: {
        recover: {
          name: ``,
          seed: ``,
          password: ``,
          passwordConfirm: ``,
          warning: false
        }
      },
      getters,
      commit: jest.fn(),
      dispatch: jest.fn(async () => true)
    }
    wrapper = mount(TmSessionImportBackupCode, {
      localVue,
      mocks: {
        $store,
        $router: {
          push: jest.fn()
        }
      },
      stubs: [`router-link`]
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`should disable button if seed is not filled in`, async () => {
    const getters = {
      connected: () => true
    }
    $store = {
      state: {
        recover: {
          name: ``,
          seed: ``,
          password: ``,
          passwordConfirm: ``,
          warning: false
        }
      },
      getters,
      commit: jest.fn(),
      dispatch: jest.fn(async () => true)
    }
    wrapper = mount(TmSessionImportBackupCode, {
      localVue,
      mocks: {
        $store,
        $router: {
          push: jest.fn()
        }
      },
      stubs: [`router-link`]
    })
    expect(wrapper.find(TmBtn).attributes(`disabled`)).toBe(`disabled`)
  })

  it(`should show error if seed is not 24 words long`, async () => {
    const getters = {
      connected: () => true,
      seed: () => `asdf asdf asdf asdf`
    }
    $store = {
      state: {
        recover: {
          name: ``,
          seed: `asdf asdf asdf asdf`,
          password: ``,
          passwordConfirm: ``,
          warning: false
        }
      },
      getters,
      commit: jest.fn(),
      dispatch: jest.fn(async () => true)
    }
    wrapper = mount(TmSessionImportBackupCode, {
      localVue,
      mocks: {
        $store,
        $router: {
          push: jest.fn()
        }
      },
      stubs: [`router-link`]
    })

    // console.log(wrapper.html())
    const input = wrapper.find(`textarea`)
    input.trigger(`input`)
    // console.log(input.html())
    expect(wrapper.find(TmBtn).attributes(`disabled`)).toBe(`disabled`)
  })

  it(`should show error if seed is not 24 words long`, async () => {
    const getters = {
      connected: () => true,
      seed: () => `asdf asdf asdf asdf`
    }
    $store = {
      state: {
        recover: {
          name: ``,
          seed: `asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf asdf`,
          password: ``,
          passwordConfirm: ``,
          warning: false
        }
      },
      getters,
      commit: jest.fn(),
      dispatch: jest.fn(async () => true)
    }
    wrapper = mount(TmSessionImportBackupCode, {
      localVue,
      mocks: {
        $store,
        $router: {
          push: jest.fn()
        }
      },
      stubs: [`router-link`]
    })
    //console.log(wrapper.html())
    expect(wrapper.find(TmBtn).attributes(`disabled`)).toBe(undefined)
  })

  /* it(`should show error if seed is not 24 words long`, async () => {
    wrapper.setData({
      fields: {
        importSeed: `asdf asdf asdf asdf`
      }
    })
    await wrapper.vm.onSubmit()
    expect($store.commit.mock.calls[0]).toBeUndefined()
    expect(wrapper.find(`.tm-form-msg-error`)).toBeDefined()
  })

  it(`should show error if password don't match`, async () => {
    wrapper.setData({
      fields: {
        importName: `foo123`,
        importPassword: `1234567890`,
        importPasswordConfirm: `notthesame`,
        importSeed: seed
      }
    })
    await wrapper.vm.onSubmit()
    expect($store.commit.mock.calls[0]).toBeUndefined()
    expect(wrapper.find(`.tm-form-msg-error`)).toBeDefined()
  })

  it(`should not continue if creation failed`, async () => {
    $store.dispatch = jest.fn(() => Promise.reject(new Error(`Wrong password`)))
    wrapper.setData({
      fields: {
        importName: `foo123`,
        importPassword: `1234567890`,
        importPasswordConfirm: `1234567890`,
        importSeed: seed
      }
    })
    await wrapper.vm.onSubmit()
    expect($store.commit).toHaveBeenCalledWith(`notifyError`, {
      title: `Couldn't create account`,
      body: expect.stringContaining(`Wrong password`)
    })
  })

  it(`should show a notification if creation failed`, async () => {
    $store.dispatch = jest.fn(() => Promise.reject({ message: `test` }))
    wrapper.setData({
      fields: {
        importName: `foo123`,
        importPassword: `1234567890`,
        importPasswordConfirm: `1234567890`,
        importSeed: seed
      }
    })
    await wrapper.vm.onSubmit()
    expect($store.commit.mock.calls[0][0]).toEqual(`notifyError`)
    expect($store.commit.mock.calls[0][1].body).toEqual(`test`)
  })

  it(`should go to the home page if recovering is successful`, async () => {
    wrapper.setData({
      fields: {
        importName: `foo123`,
        importPassword: `1234567890`,
        importPasswordConfirm: `1234567890`,
        importSeed: seed
      }
    })
    $store.dispatch = jest.fn(() => Promise.resolve())
    await wrapper.vm.onSubmit()
    expect($store.dispatch.mock.calls[0][0]).toEqual(`createKey`)
    expect(wrapper.vm.$router.push).toHaveBeenCalledWith(`/`)
  }) */
})
