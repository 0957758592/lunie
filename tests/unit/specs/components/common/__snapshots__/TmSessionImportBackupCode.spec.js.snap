// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TmSessionImportBackupCode has the expected html structure 1`] = `
<div
  class="session-frame"
  mode="out-in"
  name="component-fade"
>
  <router-link-stub
    to="/"
  >
    <img
      class="session-logo"
      src="~assets/images/lunie-logo-white.svg"
    />
  </router-link-stub>
   
  <div
    class="session-outer-container"
  >
    <div
      class="session"
    >
      <a>
        <i
          class="material-icons session-back"
        >
          arrow_back
        </i>
      </a>
       
      <form
        class="form"
      >
         
        <!---->
         
        <main
          class="form-main"
        >
          <h2
            class="session-title"
          >
            
      Recover with backup code
    
          </h2>
           
          <div
            class="session-main"
          >
            <div
              class="tm-form-group"
            >
              <!---->
               
              <label
                class="tm-form-group__label"
                for="import-seed"
              >
                
    Backup code
  
              </label>
               
              <div
                class="tm-form-group__field"
              >
                <textarea
                  class="tm-field-seed tm-field"
                  id="import-seed"
                  placeholder="Must be exactly 24 words"
                  style="overflow-x: hidden; word-wrap: break-word; overflow-y: hidden;"
                />
                 
                <!---->
              </div>
            </div>
          </div>
           
          <div
            class="session-footer"
          >
            <button
              class="button"
              disabled="disabled"
            >
              
  Confirm Seed

            </button>
          </div>
        </main>
         
        <!---->
      </form>
    </div>
  </div>
   
  <button
    class="button session-close"
    color="secondary"
  >
    
  Back to Lunie

  </button>
</div>
`;
